woocommerce/woocommerce-android
# Add support for updating card reader software

Parent issue #3957 

This PR adds support for updating card reader software.
- `CheckSoftwareUpdatesAction` and `InstallSoftwareUpdateAction` contain the actual logic for checking for updates and updating the sw
- `SoftwareUpdateManager` takes care of the orchestration -> checks for updates first and updates the sw if neccessary
- Changes in `CardReaderSettingsFragment` and layout don't need to be reviewed as the code will never be used in production - it's just a naive implementation for testing purposes

Note - Don't get scared by the size of the PR -> some of the changes don't need to be reviewed as mentioned above and over half of the line changes are unit tests. Having said that if you'd prefer me to split it into "logic", "unit tests" and "UI" PRs I can do that - I just thought it might be good to have the dummy testing UI so you can test the code.

Known issues:
1. Leaving the screen when the update is in progress results in a crash - it's reported in the parent issue and I plan to take care of it when the actual UI is implemented -> I'm not 100% sure it'll still be an issue when we start using VMs
2. Manually turning off the reader while the update is in progress sometime results in ANR dialog - I reported this issue in AndroidStripeTerminalSDK repo

To Test
- Make sure to test these changes with a site with WCPay in devmode
- Do NOT use a hw reader to test these changes as it might be a good idea to keep an old version of the software on your reader for future testing
1. Open app settings
2. Click on "Manage card readers"
3. Check "Use simulated reader" checkbox
4. Tap on "Connect Card Reader"
5. Tap on "Update Reader Software"
6. Notice the progress is updated and the update succeeds

Update release notes:
